econml/score/ensemble_cate.py

```diff
@@ -0,0 +1,68 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+from sklearn.utils.validation import check_array
+from .._cate_estimator import BaseCateEstimator, LinearCateEstimator
+
+
+class EnsembleCateEstimator:
+    """ A CATE estimator that represents a weighted ensemble of many
+    CATE estimators. Returns their weighted effect prediction.
+
+    Parameters
+    ----------
+    cate_models : list of BaseCateEstimator objects
+        A list of fitted cate estimator objects that will be used in the ensemble.
+        The models are passed by reference, and not copied internally, because we
+        need the fitted objects, so any change to the passed models will affect
+        the internal predictions (e.g. if the input models are refitted).
+    weights : np.ndarray of shape (len(cate_models),)
+        The weight placed on each model. Weights must be non-positive. The
+        ensemble will predict effects based on the weighted average predictions
+        of the cate_models estiamtors, weighted by the corresponding weight in `weights`. 
+    """
+
+    def __init__(self, *, cate_models, weights):
+        self._cate_models = cate_models
+        self._weights = weights
+
+    def effect(self, X=None, *, T0=0, T1=1):
+        return np.average([mdl.effect(X=X, T0=T0, T1=T1) for mdl in self.cate_models],
+                          weights=self.weights, axis=0)
+    effect.__doc__ = BaseCateEstimator.effect.__doc__
+
+    def marginal_effect(self, T, X=None):
+        return np.average([mdl.marginal_effect(T, X=X) for mdl in self.cate_models],
+                          weights=self.weights, axis=0)
+    marginal_effect.__doc__ = BaseCateEstimator.marginal_effect.__doc__
+
+    def const_marginal_effect(self, X=None):
+        if np.any([not hasattr(mdl, 'const_marginal_effect') for mdl in self.cate_models]):
+            raise ValueError("One of the base CATE models in parameter `cate_models` does not support "
+                             "the `const_marginal_effect` method.")
+        return np.average([mdl.const_marginal_effect(X=X) for mdl in self.cate_models],
+                          weights=self.weights, axis=0)
+    const_marginal_effect.__doc__ = LinearCateEstimator.const_marginal_effect.__doc__
+
+    @property
+    def cate_models(self):
+        return self._cate_models
+
+    @cate_models.setter
+    def cate_models(self, value):
+        if not isinstance(value, list):
+            raise ValueError('Parameter `cate_models` should be a list of `BaseCateEstimator` objects.')
+        self._cate_models = value
+
+    @property
+    def weights(self):
+        return self._weights
+
+    @weights.setter
+    def weights(self, value):
+        weights = check_array(value, accept_sparse=False, ensure_2d=False, allow_nd=False, dtype='numeric',

```

Comment:
Seems a little weird that this validation happens here, but not in the initializer.

Reply:
Why? this is a check that should happen whenever we assign a value to the weights. The initializer simply invokes the setter.

Reply:
Looks like the initializer is setting the field directly, not calling through this property.

Reply:
(and same for the models)

Reply:
I think that's done implicit. Exactly because there is a setter it automatically invokes the setter.

Reply:
```python
class Test:
    def __init__(self, value):
        self.value = 10
    
    @property
    def value(self):
        return self._value
    
    @value.setter
    def value(self, value):
        print("Test")
        self._value = value
```

Then
```
Test(10)
```
prints "Test"

Reply:
Try:

```
class Test:
    def __init__(self, value):
        self._value = value # leading underscore

    @property
    def value(self):
        return self._value

    @value.setter
    def value(self, value):
        print("Test")
        self._value = value

Reply:
(Maybe we're agreeing and I should have put my comment in the initializer instead - it needs to be setting `self.weight` and `self.cate_models`, not `self._weight` and `self._cate_models`)

Reply:
yeap! Now we're good... :) 

Fixed